### PR TITLE
📚 docs: Allow visitors to request ERDs

### PR DESCRIPTION
**Pull request by letiescanciano**
Closes https://github.com/airbytehq/airbyte-internal-issues/issues/9775
## What
- We think it would be interesting to allow visitors to request ERDs for sources.
Screenshot 2024-09-23 at 17 04 55

## How
- Create new <RequestERD /> component 
- Create <EmailModal  /> component, we want to get  the visitor's email to be able to notify them when/if the ERD is ready.
- Request will be submitted to our internal API and save in Retool DB.

## User Impact
- Users will be able to request ERDs for sources from the TOC.

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: docusaurus/src/theme/TOCItems/index.js [hunk 1]
```diff
@@ -1,22 +1,54 @@
-import React, { useState } from 'react';
-import TOCItems from '@theme-original/TOCItems';
-import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
-import { faThumbsDown, faThumbsUp } from '@fortawesome/free-regular-svg-icons';
+import { faThumbsDown, faThumbsUp } from "@fortawesome/free-regular-svg-icons";
```
Comment (letiescanciano) [inline docusaurus/src/theme/TOCItems/index.js:1]: No code changes in this file, just formatting

File: docusaurus/src/components/RequestERD/EmailModal.jsx [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+          disabled={status === "loading"}
+        />
+        {emailError && <p className={styles.inputErrorMessage}>{emailError}</p>}
+      </div>
+      <button
+        type="submit"
+        onClick={(e) => {
+          e.preventDefault();
+          onSubmit(email);
+        }}
```
Comment (lmossman) [inline docusaurus/src/components/RequestERD/EmailModal.jsx:74]: Why is this necessary? From my understanding, having a button with `type="submit"` in a form should trigger the form's `onSubmit` callback, which is what this onClick handler is doing anyway?

File: docusaurus/src/components/RequestERD/index.jsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import { useLocation } from "@docusaurus/router";
+import React, { useEffect, useState } from "react";
+import { EmailModal } from "./EmailModal";
+import styles from "./RequestERD.module.css";
+import { faPlus } from "@fortawesome/free-solid-svg-icons";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+const { getRegistryEntry } = require("../../remark/utils");
+const { getFromPaths } = require("../../helpers/objects");
+
+export const RequestERD = () => {
```
Comment (lmossman) [inline docusaurus/src/components/RequestERD/index.jsx:10]: Let's rename this file to `RequestERD.jsx` to be consistent with other components in this repo

File: docusaurus/src/components/RequestERD/EmailModal.jsx [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+          autoComplete="email"
+          onChange={handleOnChange}
+          placeholder="Enter your email"
+          required
+          disabled={status === "loading"}
+        />
+        {emailError && <p className={styles.inputErrorMessage}>{emailError}</p>}
+      </div>
+      <button
+        type="submit"
```
Comment (lmossman) [inline docusaurus/src/components/RequestERD/EmailModal.jsx:70]: The submit button and the input are different heights - let's make them the same height
Screenshot 2024-10-08 at 3 22 46 PM

File: docusaurus/src/components/RequestERD/EmailModal.jsx [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  const [emailError, setEmailError] = useState("");
+
+  const debouncedValidation = useDebouncedCallback((value) => {
+    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
+    if (emailRegex.test(value)) {
+      setEmailError("");
+    } else {
+      setEmailError("Please enter a valid email address");
+    }
+  }, 800);
```
Comment (lmossman) [inline docusaurus/src/components/RequestERD/EmailModal.jsx:28]: I don't think a debounce provides the best UX here, as it makes the validation feel laggy.

Instead, I think a better UX would be:
- Perform the validation on every change to the input
- Only show the error if the user blurs the input while there is a value in it. This way we give them a chance to write a full without showing any errors, even if they pause between some of the characters, and only show the error if they then leave the input (e.g. by clicking the Submit button)

File: docusaurus/src/components/RequestERD/EmailModal.jsx [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        {emailError && <p className={styles.inputErrorMessage}>{emailError}</p>}
+      </div>
+      <button
+        type="submit"
+        onClick={(e) => {
+          e.preventDefault();
+          onSubmit(email);
+        }}
+        className={styles.modalContentButton}
+        disabled={status === "loading"}
```
Comment (lmossman) [inline docusaurus/src/components/RequestERD/EmailModal.jsx:76]: We should also disable this button if the email input is in an error state

File: docusaurus/src/components/RequestERD/EmailModal.jsx [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      </div>
+    );
+  }
+  if (status === "error") {
+    return (
+      <div>
+        <p> Error submitting request. Please try again later.</p>
+      </div>
+    );
+  }
```
Comment (lmossman) [inline docusaurus/src/components/RequestERD/EmailModal.jsx:50]: When I submitted with an invalid email, the modal showed me this error, and even after closing the modal and clicking the `+ Request Entity-Relationship Diagram` button again, the modal opened in this error state rather than letting me input a new email and try again.

I could only get it out of this state by refreshing the page.

File: docusaurus/src/components/RequestERD/EmailModal.jsx [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+  return (
+    <form onSubmit={() => onSubmit(email)} className={styles.modalContentForm}>
+      <div className={styles.modalContentInputContainer}>
+        <input
+          className={cn(styles.modalContentInput, {
+            [styles.modalContentInputError]: emailError,
+          })}
```
Comment (lmossman) [inline docusaurus/src/components/RequestERD/EmailModal.jsx:58]: It would look more consistent with the rest of our UI if we changed the border of the input to red, and kept the text color the same, when this input is in an error state.

File: docusaurus/src/components/RequestERD/EmailModal.jsx [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+  const {
+    siteConfig: {
+      customFields: { requestErdApiUrl },
+    },
+  } = useDocusaurusContext();
+ 
+
+  if (!isOpen) return null;
+
+  const handleSubmit = async (email) => {
```
Comment (lmossman) [inline docusaurus/src/components/RequestERD/EmailModal.jsx:95]: I always get an error when submitting my email, even if my email is valid.

File: docusaurus/src/components/RequestERD/EmailModal.jsx [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        className={styles.modalContentButton}
+        disabled={status === "loading"}
+      >
+        {status === "loading" ? "Submitting..." : "Submit"}
+      </button>
+    </form>
+  );
+};
+
+export const EmailModal = ({ isOpen, onClose, sourceInfo }) => {
```
Comment (lmossman) [inline docusaurus/src/components/RequestERD/EmailModal.jsx:84]: It would be helpful to explain to users why we are asking them for their email. 

Maybe we could include a message that says something like "Please provide your email to request an ERD for this connector, and we will notify you when it is available" (or maybe something more concise)

cc @nataliekwong in case you have thoughts here

**Review by lmossman — CHANGES_REQUESTED** [review]

File: docusaurus/src/components/RequestERD/EmailModal.jsx (hunk 2 above)
Comment (letiescanciano) [reply docusaurus/src/components/RequestERD/EmailModal.jsx:74]: You are absolutely right. The issue I was having was that submitting the form was causing the browser to reload the page.
With `e.preventDefault()` I can then handle the submission and stay in the same page.
Thinking about it again, I think a better place for this is on the `onSubmit` method on` <form>` instead of on the button itself. What do you think?

```
   <form onSubmit={(e) => {
      e.preventDefault();
      onSubmit(email);
    }} >
 ```

File: docusaurus/src/components/RequestERD/EmailModal.jsx (hunk 5 above)
Comment (letiescanciano) [reply docusaurus/src/components/RequestERD/EmailModal.jsx:28]: Great suggestion. Definitely better UX

File: docusaurus/src/components/RequestERD/EmailModal.jsx (hunk 7 above)
Comment (letiescanciano) [reply docusaurus/src/components/RequestERD/EmailModal.jsx:50]: Nice catch. I'm resetting `status` after closing the modal

File: docusaurus/src/components/RequestERD/EmailModal.jsx (hunk 9 above)
Comment (letiescanciano) [reply docusaurus/src/components/RequestERD/EmailModal.jsx:113]: I’m sorry, I should have explained how to test this in the PR.
This submission is sent to an api endpoint in the `airbyte-web-experiments` project, which is also hosted in vercel.
The [endpoint](https://github.com/airbytehq/airbyte-web-experiments/blob/935248635f1713d23cd2557fdcb41c9f3615a6eb/pages/api/request-erd/index.ts#L6) will take care of triggering the Retool workflow.
Since origins are different for both projects, I had to set CORS at the [middleware](https://github.com/airbytehq/airbyte-web-experiments/blob/935248635f1713d23cd2557fdcb41c9f3615a6eb/middleware.ts#L7) 
Anyway, I had an error on the env variables that make this possible.

If you test it in the [preview](https://airbyte-docs-git-leti-allow-request-erds-on-docs-airbyte-growth.vercel.app/integrations/sources/activecampaign) should work now.

File: docusaurus/src/components/RequestERD/EmailModal.module.css [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  color: white;
+  border: none;
+  padding: 10px 20px;
+  border-radius: 6px;
+  cursor: pointer;
+  max-height: 38px;
+}
+
+.modalContentButton:disabled {
+  background-color: var(--ifm-color-primary-light);
```
Comment (lmossman) [inline docusaurus/src/components/RequestERD/EmailModal.module.css:75]: In the airbyte webapp, we usually just set the opacity to 50% for disabled buttons - I think this current color looks a little bit too close to the non-disabled color, so it wasn't as clear to me that it was disabled. 

Setting opacity to 50% instead would probably make this distinction clearer

File: docusaurus/src/components/RequestERD/EmailModal.jsx [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+      } else {
+        setStatus("success");
+      }
+    } catch (error) {
+      console.error("Error requesting ERD:", error);
+      setStatus("error");
+    }
+  };
+  return (
+    <div className={styles.modalOverlay}>
```
Comment (lmossman) [inline docusaurus/src/components/RequestERD/EmailModal.jsx:143]: Currently, clicking this overlay background does not close the modal, which I expected it to as that is pretty standard modal behavior.

If you used the [Headless UI Modal](https://headlessui.com/v1/react/dialog) component here, then you would get this behavior (and many others) out of the box, so you wouldn't need to implement it yourself. This is the approach I would recommend

**Review by lmossman — APPROVED** [review]
Had a few more small comments, but I don't think anything is truly blocking and can probably be addressed without a re-review from me.

But if you want me to take another look feel free to re-request me!
